# Data Project: SQL Database 
This file is used to clean the fitbit_fitness_tracker_RAW.csv file

# Importing Libraries

In [352]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from IPython.display import display
from datetime import datetime
warnings.filterwarnings('ignore')

# Loading the Fitbit CSV Data 

In [353]:
# Definig the file path 
file_path = "Datasets_files/fitbit_fitness_tracker_RAW.csv"

# Loading the CSV file into a dataframe 
fitbit_df = pd.read_csv(file_path)

# Displaying the first 10 rows of the dataframe
fitbit_df.head(10)

Id               SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
0  1503960366  4/12/2016 12:00:00 AM                  1                 327   
1  1503960366  4/12/2016 12:00:00 AM                  1                 327   
2  1503960366  4/12/2016 12:00:00 AM                  1                 327   
3  1503960366  4/12/2016 12:00:00 AM                  1                 327   
4  1503960366  4/12/2016 12:00:00 AM                  1                 327   
5  1503960366  4/12/2016 12:00:00 AM                  1                 327   
6  1503960366  4/12/2016 12:00:00 AM                  1                 327   
7  1503960366  4/12/2016 12:00:00 AM                  1                 327   
8  1503960366  4/12/2016 12:00:00 AM                  1                 327   
9  1503960366  4/12/2016 12:00:00 AM                  1                 327   

   TotalTimeInBed ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0             346     5/7/2016       11992           7.71             7.71   
1             346     5/6/2016       12159           8.03             8.03   
2             346     5/1/2016       10602           6.81             6.81   
3             346    4/30/2016       14673           9.25             9.25   
4             346    4/12/2016       13162           8.50             8.50   
5             346    4/13/2016       10735           6.97             6.97   
6             346    4/25/2016       15355           9.80             9.80   
7             346     5/5/2016       14070           8.90             8.90   
8             346    4/28/2016       13154           8.53             8.53   
9             346    4/29/2016       11181           7.15             7.15   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.46                      2.12   
1                       0.0                1.97                      0.25   
2                       0.0                2.29                      1.60   
3                       0.0                3.56                      1.42   
4                       0.0                1.88                      0.55   
5                       0.0                1.57                      0.69   
6                       0.0                5.29                      0.57   
7                       0.0                2.92                      1.08   
8                       0.0                3.54                      1.16   
9                       0.0                1.06                      0.50   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 3.13                      0.0                 37   
1                 5.81                      0.0                 24   
2                 2.92                      0.0                 33   
3                 4.27                      0.0                 52   
4                 6.06                      0.0                 25   
5                 4.71                      0.0                 21   
6                 3.94                      0.0                 73   
7                 4.88                      0.0                 45   
8                 3.79                      0.0                 48   
9                 5.58                      0.0                 16   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   46                   175               833      1821  
1                    6                   289               754      1896  
2                   35                   246               730      1820  
3                   34                   217               712      1947  
4                   13                   328               728      1985  
5                   19                   217               776      1797  
6                   14                   216               814      2013  
7                   24                   250               857 

In [354]:
# Reviewing the names of the columns in a list form

#print(fitbit_df.columns.tolist())
for i, col in enumerate(fitbit_df.columns):
    print(f"{i}: {col}")

0: Id
1: SleepDay
2: TotalSleepRecords
3: TotalMinutesAsleep
4: TotalTimeInBed
5: ActivityDate
6: TotalSteps
7: TotalDistance
8: TrackerDistance
9: LoggedActivitiesDistance
10: VeryActiveDistance
11: ModeratelyActiveDistance
12: LightActiveDistance
13: SedentaryActiveDistance
14: VeryActiveMinutes
15: FairlyActiveMinutes
16: LightlyActiveMinutes
17: SedentaryMinutes
18: Calories


# Reviewing the dataset for any null, missing values, and duplicates.

In [355]:
# Obtaining and Displaying the missing data points per column 
missing_values_count = fitbit_df.isnull().sum()
missing_values_count[:]

Id                          0
SleepDay                    0
TotalSleepRecords           0
TotalMinutesAsleep          0
TotalTimeInBed              0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64

In [356]:
# Checking for duplicates 
fitbit_df.duplicated().sum()

93

In [357]:
# Showing where the duplicates are
fitbit_df[fitbit_df.duplicated()]

Id               SleepDay  TotalSleepRecords  \
4764   4388161847   5/5/2016 12:00:00 AM                  1   
4765   4388161847   5/5/2016 12:00:00 AM                  1   
4766   4388161847   5/5/2016 12:00:00 AM                  1   
4767   4388161847   5/5/2016 12:00:00 AM                  1   
4768   4388161847   5/5/2016 12:00:00 AM                  1   
...           ...                    ...                ...   
11474  8378563200  4/25/2016 12:00:00 AM                  1   
11475  8378563200  4/25/2016 12:00:00 AM                  1   
11476  8378563200  4/25/2016 12:00:00 AM                  1   
11477  8378563200  4/25/2016 12:00:00 AM                  1   
11478  8378563200  4/25/2016 12:00:00 AM                  1   

       TotalMinutesAsleep  TotalTimeInBed ActivityDate  TotalSteps  \
4764                  471             495    4/17/2016        4660   
4765                  471             495    4/18/2016       11009   
4766                  471             495    4/19/2016       10181   
4767                  471             495    4/21/2016       10055   
4768                  471             495    4/15/2016        8758   
...                   ...             ...          ...         ...   
11474                 388             402     5/3/2016        8712   
11475                 388             402    4/25/2016       12405   
11476                 388             402    4/26/2016       16208   
11477                 388             402    4/23/2016        5709   
11478                 388             402    4/24/2016        3703   

       TotalDistance  TrackerDistance  LoggedActivitiesDistance  \
4764            3.58             3.58                  0.000000   
4765            9.10             9.10                  0.000000   
4766            7.83             7.83                  0.000000   
4767            7.73             7.73                  0.000000   
4768            6.73             6.73                  0.000000   
...              ...              ...                       ...   
11474           6.91             6.91                  2.253081   
11475           9.84             9.84                  2.092147   
11476          12.85            12.85                  0.000000   
11477           4.53             4.53                  0.000000   
11478           2.94             2.94                  0.000000   

       VeryActiveDistance  ModeratelyActiveDistance  LightActiveDistance  \
4764                 0.00                      0.00                 3.58   
4765                 3.56                      0.40                 5.14   
4766                 1.37                      0.69                 5.77   
4767                 0.37                      0.39                 6.98   
4768                 0.00                      0.00                 6.73   
...                   ...                       ...                  ...   
11474                1.34                      1.06                 4.50   
11475                5.05                      0.87                 3.92   
11476                7.51                      0.92                 4.42   
11477                1.52                      0.52                 2.48   
11478                0.00                      0.00                 2.94   

       SedentaryActiveDistance  VeryActiveMinutes  FairlyActiveMinutes  \
4764                       0.0                  0                    0   
4765                       0.0                 27                    8   
4766                       0.0                 20                   16   
4767                       0.0                  7                   12   
4768                       0.0                  0                    0   
...                        ...                ...                  ...   
11474                      0.0                 71                   20   
11475                      0.0                117                   16   
11476                      0.0   

In [358]:
fitbit_df.duplicated(subset=["SleepDay", "ActivityDate"]).sum()
fitbit_df[fitbit_df.duplicated(subset=["SleepDay", "ActivityDate"], keep=False)]

Id               SleepDay  TotalSleepRecords  \
0      1503960366  4/12/2016 12:00:00 AM                  1   
1      1503960366  4/12/2016 12:00:00 AM                  1   
2      1503960366  4/12/2016 12:00:00 AM                  1   
3      1503960366  4/12/2016 12:00:00 AM                  1   
4      1503960366  4/12/2016 12:00:00 AM                  1   
...           ...                    ...                ...   
12436  8792009665   5/4/2016 12:00:00 AM                  1   
12437  8792009665   5/4/2016 12:00:00 AM                  1   
12438  8792009665   5/4/2016 12:00:00 AM                  1   
12439  8792009665   5/4/2016 12:00:00 AM                  1   
12440  8792009665   5/4/2016 12:00:00 AM                  1   

       TotalMinutesAsleep  TotalTimeInBed ActivityDate  TotalSteps  \
0                     327             346     5/7/2016       11992   
1                     327             346     5/6/2016       12159   
2                     327             346     5/1/2016       10602   
3                     327             346    4/30/2016       14673   
4                     327             346    4/12/2016       13162   
...                   ...             ...          ...         ...   
12436                 439             463    4/27/2016        1758   
12437                 439             463    4/28/2016        6157   
12438                 439             463    4/29/2016        8360   
12439                 439             463    4/30/2016        7174   
12440                 439             463     5/1/2016        1619   

       TotalDistance  TrackerDistance  LoggedActivitiesDistance  \
0               7.71             7.71                       0.0   
1               8.03             8.03                       0.0   
2               6.81             6.81                       0.0   
3               9.25             9.25                       0.0   
4               8.50             8.50                       0.0   
...              ...              ...                       ...   
12436           1.13             1.13                       0.0   
12437           3.94             3.94                       0.0   
12438           5.35             5.35                       0.0   
12439           4.59             4.59                       0.0   
12440           1.04             1.04                       0.0   

       VeryActiveDistance  ModeratelyActiveDistance  LightActiveDistance  \
0                    2.46                      2.12                 3.13   
1                    1.97                      0.25                 5.81   
2                    2.29                      1.60                 2.92   
3                    3.56                      1.42                 4.27   
4                    1.88                      0.55                 6.06   
...                   ...                       ...                  ...   
12436                0.00                      0.00                 1.13   
12437                0.00                      0.00                 3.94   
12438                0.14                      0.28                 4.93   
12439                0.33                      0.36                 3.91   
12440                0.00                      0.00                 1.04   

       SedentaryActiveDistance  VeryActiveMinutes  FairlyActiveMinutes  \
0                          0.0                 37                   46   
1                          0.0                 24                    6   
2                          0.0                 33                   35   
3                          0.0                 52                   34   
4                          0.0                 25                   13   
...                        ...                ...                  ...   
12436                      0.0                  0                    0   
12437                      0.0                  0                    0   
12438                      0.0   

**Finding out the basic information of the fitbit dataset:**
- no. of rows and columns
- name of columns 
- type of values

In [359]:
# Showing the basic information of the dataset
fitbit_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12441 entries, 0 to 12440
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        12441 non-null  int64  
 1   SleepDay                  12441 non-null  object 
 2   TotalSleepRecords         12441 non-null  int64  
 3   TotalMinutesAsleep        12441 non-null  int64  
 4   TotalTimeInBed            12441 non-null  int64  
 5   ActivityDate              12441 non-null  object 
 6   TotalSteps                12441 non-null  int64  
 7   TotalDistance             12441 non-null  float64
 8   TrackerDistance           12441 non-null  float64
 9   LoggedActivitiesDistance  12441 non-null  float64
 10  VeryActiveDistance        12441 non-null  float64
 11  ModeratelyActiveDistance  12441 non-null  float64
 12  LightActiveDistance       12441 non-null  float64
 13  SedentaryActiveDistance   12441 non-null  float64
 14  VeryAc

In [360]:
# Counting the unique ID and reviewing how much IDs it has
#fitbit_df["Id"].nunique()
unique_id = len(pd.unique(fitbit_df["Id"])) 
print("Amount of unique Id: " + str(unique_id))

Amount of unique Id: 24


In [361]:
# Checking for thhe amount of rows & columns
#print(len(fitbit_df))
print(fitbit_df.shape)

(12441, 19)


**Based on the initial observations, I conclude the following:**
There are no missing or null values in the dataset.
The dataset contains 19 columns and 12,441 rows.
The ActivityDate column is currently of type object and should be converted to datetime64.
There are 24 unique user IDs in the dataset.
There are 93 duplicated rows based on the Id column. However, these duplicates differ in key activity-related columns such as ActivityDate, TotalSteps, TotalDistance, TrackerDistance, LoggedActivitiesDistance, VeryActiveDistance, ModeratelyActiveDistance, LightActiveDistance, SedentaryActiveDistance, VeryActiveMinutes, FairlyActiveMinutes, LightlyActiveMinutes, SedentaryMinutes, and Calories, indicating that they represent different records for the same user.

**The following data manipulations were performed:**
- Converted the ActivityDate column to datetime64 data type.
- Reformatted ActivityDate to follow the YYYY-MM-DD format.
- Created a new column, DayOfTheWeek, to extract the day name from each date for further analysis.
- Added a new column, TotalMins, representing the total of VeryActiveMinutes, FairlyActiveMinutes, LightlyActiveMinutes, and SedentaryMinutes.
- Created a new column, TotalHours, by converting TotalMins into hours.
- Renamed and rearranged columns for improved readability and structure.

In [362]:
# Converted the ActivityDate column to datetime64 data type and formatted to DD-MM-YY format
#fitbit_df["ActivityDate"] = pd.to_datetime(fitbit_df["ActivityDate"], format="%d/%m/%Y")

# Converting the ActivityDate column to datatime64 dtype and format to dd-mm-yyyy
# Note: data might seems off to parse it to EU format, so it will be parsed to US format
fitbit_df["ActivityDate"] = pd.to_datetime(fitbit_df["ActivityDate"], format="%m/%d/%Y")
fitbit_df.info()
fitbit_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12441 entries, 0 to 12440
Data columns (total 19 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        12441 non-null  int64         
 1   SleepDay                  12441 non-null  object        
 2   TotalSleepRecords         12441 non-null  int64         
 3   TotalMinutesAsleep        12441 non-null  int64         
 4   TotalTimeInBed            12441 non-null  int64         
 5   ActivityDate              12441 non-null  datetime64[ns]
 6   TotalSteps                12441 non-null  int64         
 7   TotalDistance             12441 non-null  float64       
 8   TrackerDistance           12441 non-null  float64       
 9   LoggedActivitiesDistance  12441 non-null  float64       
 10  VeryActiveDistance        12441 non-null  float64       
 11  ModeratelyActiveDistance  12441 non-null  float64       
 12  LightActiveDistanc

Id               SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
0  1503960366  4/12/2016 12:00:00 AM                  1                 327   
1  1503960366  4/12/2016 12:00:00 AM                  1                 327   
2  1503960366  4/12/2016 12:00:00 AM                  1                 327   
3  1503960366  4/12/2016 12:00:00 AM                  1                 327   
4  1503960366  4/12/2016 12:00:00 AM                  1                 327   

   TotalTimeInBed ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0             346   2016-05-07       11992           7.71             7.71   
1             346   2016-05-06       12159           8.03             8.03   
2             346   2016-05-01       10602           6.81             6.81   
3             346   2016-04-30       14673           9.25             9.25   
4             346   2016-04-12       13162           8.50             8.50   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                2.46                      2.12   
1                       0.0                1.97                      0.25   
2                       0.0                2.29                      1.60   
3                       0.0                3.56                      1.42   
4                       0.0                1.88                      0.55   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 3.13                      0.0                 37   
1                 5.81                      0.0                 24   
2                 2.92                      0.0                 33   
3                 4.27                      0.0                 52   
4                 6.06                      0.0                 25   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   46                   175               833      1821  
1                    6                   289               754      1896  
2                   35                   246               730      1820  
3                   34                   217               712      1947  
4                   13                   328               728      1985

In [363]:
print(fitbit_df["ActivityDate"].head(10))

0   2016-05-07
1   2016-05-06
2   2016-05-01
3   2016-04-30
4   2016-04-12
5   2016-04-13
6   2016-04-25
7   2016-05-05
8   2016-04-28
9   2016-04-29
Name: ActivityDate, dtype: datetime64[ns]


In [364]:
# Dataset with new arranged columns
new_columns = ['Id', 'ActivityDate', 'DayOfTheWeek', 'TotalSteps', 'TotalDistance', 'TrackerDistance', 'LoggedActivitiesDistance', 'VeryActiveDistance', 'ModeratelyActiveDistance', 'LightActiveDistance', 'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes', 'LightlyActiveMinutes', 'SedentaryMinutes', 'TotalExerciseMinutes', 'TotalExerciseHours', 'Calories']
new_fitbit_df = fitbit_df.reindex(columns=new_columns)
new_fitbit_df.head()

Id ActivityDate  DayOfTheWeek  TotalSteps  TotalDistance  \
0  1503960366   2016-05-07           NaN       11992           7.71   
1  1503960366   2016-05-06           NaN       12159           8.03   
2  1503960366   2016-05-01           NaN       10602           6.81   
3  1503960366   2016-04-30           NaN       14673           9.25   
4  1503960366   2016-04-12           NaN       13162           8.50   

   TrackerDistance  LoggedActivitiesDistance  VeryActiveDistance  \
0             7.71                       0.0                2.46   
1             8.03                       0.0                1.97   
2             6.81                       0.0                2.29   
3             9.25                       0.0                3.56   
4             8.50                       0.0                1.88   

   ModeratelyActiveDistance  LightActiveDistance  SedentaryActiveDistance  \
0                      2.12                 3.13                      0.0   
1                      0.25                 5.81                      0.0   
2                      1.60                 2.92                      0.0   
3                      1.42                 4.27                      0.0   
4                      0.55                 6.06                      0.0   

   VeryActiveMinutes  FairlyActiveMinutes  LightlyActiveMinutes  \
0                 37                   46                   175   
1                 24                    6                   289   
2                 33                   35                   246   
3                 52                   34                   217   
4                 25                   13                   328   

   SedentaryMinutes  TotalExerciseMinutes  TotalExerciseHours  Calories  
0               833                   NaN                 NaN      1821  
1               754                   NaN                 NaN      1896  
2               730                   NaN                 NaN      1820  
3               712                   NaN                 NaN      1947  
4               728                   NaN                 NaN      1985

In [365]:
# Creating a new column for the day of the week 
new_fitbit_df["DayOfTheWeek"] = new_fitbit_df["ActivityDate"].dt.day_name()

#printing the first 10 rows for DayOfTheWeek column
new_fitbit_df["DayOfTheWeek"].head(10)

0     Saturday
1       Friday
2       Sunday
3     Saturday
4      Tuesday
5    Wednesday
6       Monday
7     Thursday
8     Thursday
9       Friday
Name: DayOfTheWeek, dtype: object

In [366]:
print(new_fitbit_df.columns)

Index(['Id', 'ActivityDate', 'DayOfTheWeek', 'TotalSteps', 'TotalDistance',
       'TrackerDistance', 'LoggedActivitiesDistance', 'VeryActiveDistance',
       'ModeratelyActiveDistance', 'LightActiveDistance',
       'SedentaryActiveDistance', 'VeryActiveMinutes', 'FairlyActiveMinutes',
       'LightlyActiveMinutes', 'SedentaryMinutes', 'TotalExerciseMinutes',
       'TotalExerciseHours', 'Calories'],
      dtype='object')


In [367]:
fitbit_df = new_fitbit_df.rename(columns= {
    'Id': 'id',
    'ActivityDate': 'activity_date',
    'DayOfTheWeek': 'day_of_the_week',
    'TotalSteps': 'total_steps',
    'TotalDistance': 'total_distance',
    'TrackerDistance': 'tracker_distance',
    'LoggedActivitiesDistance': 'logged_activities_distance',
    'VeryActiveDistance': 'very_active_distance',
    'ModeratelyActiveDistance': 'moderately_active_distance',
    'LightActiveDistance': 'light_active_distance',
    'SedentaryActiveDistance': 'sedentary_active_distance',
    'VeryActiveMinutes': 'very_active_minutes',
    'FairlyActiveMinutes': 'fairly_active_minutes',
    'LightlyActiveMinutes': 'lightly_active_minutes',
    'SedentaryMinutes': 'sedentary_minutes',
    'TotalExerciseMinutes': 'total_exercise_minutes',
    'TotalExerciseHours': 'total_exercise_hours',
    'Calories': 'calories'
})

In [368]:
fitbit_df.head()

id activity_date day_of_the_week  total_steps  total_distance  \
0  1503960366    2016-05-07        Saturday        11992            7.71   
1  1503960366    2016-05-06          Friday        12159            8.03   
2  1503960366    2016-05-01          Sunday        10602            6.81   
3  1503960366    2016-04-30        Saturday        14673            9.25   
4  1503960366    2016-04-12         Tuesday        13162            8.50   

   tracker_distance  logged_activities_distance  very_active_distance  \
0              7.71                         0.0                  2.46   
1              8.03                         0.0                  1.97   
2              6.81                         0.0                  2.29   
3              9.25                         0.0                  3.56   
4              8.50                         0.0                  1.88   

   moderately_active_distance  light_active_distance  \
0                        2.12                   3.13   
1                        0.25                   5.81   
2                        1.60                   2.92   
3                        1.42                   4.27   
4                        0.55                   6.06   

   sedentary_active_distance  very_active_minutes  fairly_active_minutes  \
0                        0.0                   37                     46   
1                        0.0                   24                      6   
2                        0.0                   33                     35   
3                        0.0                   52                     34   
4                        0.0                   25                     13   

   lightly_active_minutes  sedentary_minutes  total_exercise_minutes  \
0                     175                833                     NaN   
1                     289                754                     NaN   
2                     246                730                     NaN   
3                     217                712                     NaN   
4                     328                728                     NaN   

   total_exercise_hours  calories  
0                   NaN      1821  
1                   NaN      1896  
2                   NaN      1820  
3                   NaN      1947  
4                   NaN      1985

In [369]:
# creating a new column "total_minutes" containing sum of total minutes.
fitbit_df["total_exercise_minutes"] = fitbit_df["very_active_minutes"] + fitbit_df["fairly_active_minutes"] + fitbit_df["lightly_active_minutes"] + fitbit_df["sedentary_minutes"]
fitbit_df["total_exercise_minutes"].head(5)

0    1091
1    1073
2    1044
3    1015
4    1094
Name: total_exercise_minutes, dtype: int64

In [370]:
#this would convert it into hours and minutes
# fitbit_df["total_time"] = fitbit_df["total_minutes"].apply(
#     lambda x: pd.Timedelta(minutes=x).components
# )

# fitbit_df["total_time"] = fitbit_df["total_minutes"].apply(
#     lambda x: f"{pd.Timedelta(minutes=x).components.hours:02d}:{pd.Timedelta(minutes=x).components.minutes:02d}"
# )
# print(fitbit_df["total_time"])

In [371]:
fitbit_df["total_exercise_hours"] = round(fitbit_df["total_exercise_minutes"] / 60)

In [372]:
fitbit_df.head()

id activity_date day_of_the_week  total_steps  total_distance  \
0  1503960366    2016-05-07        Saturday        11992            7.71   
1  1503960366    2016-05-06          Friday        12159            8.03   
2  1503960366    2016-05-01          Sunday        10602            6.81   
3  1503960366    2016-04-30        Saturday        14673            9.25   
4  1503960366    2016-04-12         Tuesday        13162            8.50   

   tracker_distance  logged_activities_distance  very_active_distance  \
0              7.71                         0.0                  2.46   
1              8.03                         0.0                  1.97   
2              6.81                         0.0                  2.29   
3              9.25                         0.0                  3.56   
4              8.50                         0.0                  1.88   

   moderately_active_distance  light_active_distance  \
0                        2.12                   3.13   
1                        0.25                   5.81   
2                        1.60                   2.92   
3                        1.42                   4.27   
4                        0.55                   6.06   

   sedentary_active_distance  very_active_minutes  fairly_active_minutes  \
0                        0.0                   37                     46   
1                        0.0                   24                      6   
2                        0.0                   33                     35   
3                        0.0                   52                     34   
4                        0.0                   25                     13   

   lightly_active_minutes  sedentary_minutes  total_exercise_minutes  \
0                     175                833                    1091   
1                     289                754                    1073   
2                     246                730                    1044   
3                     217                712                    1015   
4                     328                728                    1094   

   total_exercise_hours  calories  
0                  18.0      1821  
1                  18.0      1896  
2                  17.0      1820  
3                  17.0      1947  
4                  18.0      1985

## Further Analysis 

In [ ]:
# reviewing the general statistics of the dataframe
fitbit_df.describe()

id                  activity_date   total_steps  \
count  1.244100e+04                          12441  12441.000000   
mean   5.027013e+09  2016-04-26 16:00:41.668676352   8117.309220   
min    1.503960e+09            2016-04-12 00:00:00      0.000000   
25%    3.977334e+09            2016-04-19 00:00:00   4660.000000   
50%    4.702922e+09            2016-04-27 00:00:00   8596.000000   
75%    6.962181e+09            2016-05-04 00:00:00  11317.000000   
max    8.792010e+09            2016-05-12 00:00:00  22988.000000   
std    2.047809e+09                            NaN   4478.635437   

       total_distance  tracker_distance  logged_activities_distance  \
count    12441.000000      12441.000000                12441.000000   
mean         5.734738          5.727768                    0.123330   
min          0.000000          0.000000                    0.000000   
25%          3.180000          3.180000                    0.000000   
50%          6.120000          6.120000                    0.000000   
75%          7.920000          7.890000                    0.000000   
max         17.950001         17.950001                    4.942142   
std          3.243540          3.235954                    0.553071   

       very_active_distance  moderately_active_distance  \
count          12441.000000                12441.000000   
mean               1.398613                    0.732190   
min                0.000000                    0.000000   
25%                0.000000                    0.000000   
50%                0.530000                    0.400000   
75%                2.310000                    1.000000   
max               13.400000                    6.480000   
std                1.912730                    1.037477   

       light_active_distance  sedentary_active_distance  very_active_minutes  \
count           12441.000000               12441.000000         12441.000000   
mean                3.541771                   0.000674            23.973555   
min                 0.000000                   0.000000             0.000000   
25%                 2.370000                   0.000000             0.000000   
50%                 3.540000                   0.000000             8.000000   
75%                 4.830000                   0.000000            36.000000   
max                10.300000                   0.110000           210.000000   
std                 1.878553                   0.007813            34.911291   

       fairly_active_minutes  lightly_active_minutes  sedentary_minutes  \
count           12441.000000            12441.000000       12441.000000   
mean               17.352222              199.907081         799.194759   
min                 0.000000                0.000000           0.000000   
25%                 0.000000              144.000000         659.000000   
50%                10.000000              200.000000         734.000000   
75%                24.000000              258.000000         853.000000   
max               143.000000              518.000000        1440.000000   
std                23.020340               97.241286         266.778654   

       total_exercise_minutes  total_exercise_hours      calories  
count            12441.000000           12441.00000  12441.000000  
mean              1040.427618              17.33679   2329.143156  
min                  2.000000               0.00000      0.000000  
25%                927.000000              15.00000   1783.000000  
50%               1002.000000              17.00000   2162.000000  
75%               1102.000000              18.00000   2865.000000  
max               1440.000000              24.00000   4900.000000  
std                232.929705               3.89330    762.027131

**Observations**
1. Total Sample:
- The dataset includes 12,441 activity records, each representing a user’s activity data for a day.

2. Steps & Distance:
- The average user walks ~8,117 steps/day, covering ~5.73 km.
- The maximum recorded steps in a day is 22,988.

3. Activity Minutes:
- Most minutes are spent in sedentary activity (median: 734 minutes, or ~12.2 hours).
- Very active minutes have a wide range (0 to 210), with a low median of 8 minutes — suggesting low-intensity activity for most users.
- Total exercise time (very + fairly + lightly active minutes) ranges from 2 to 1,440 minutes.

4. Calories Burned:
- The average daily calories burned is ~2,329 kcal, ranging from 0 to 4,900.
- There is high variance in calorie expenditure, reflecting differences in user lifestyle or data completeness.

5. Data Quality:
- Some entries have zero values across all activity columns, which might indicate missing or non-tracked days.
- activity_date is well-distributed between April 12 and May 12, 2016, spanning one month.**